### PR TITLE
Give `Merge` a `RefCounter`

### PR DESCRIPTION
**Pull request by jorisdral**
This is WIP

### CODE REVIEW

File: src/Database/LSMTree/Internal/Merge.hs [hunk 1]
```diff
@@ -83,27 +84,44 @@ new fs hbio mergeCaching alloc mergeLevel mergeMappend targetPaths runs = do
       let numEntries = coerce (sum @V.Vector @Int) (fmap Run.runNumEntries runs)
       mergeBuilder <- Builder.new fs targetPaths numEntries alloc
       mergeLastStepResult <- newMutVar $! MergeInProgress
+      mergeRefCounter <-
+        RC.unsafeMkRefCounterN (RefCount 1)
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/Merge.hs:88]: Use `RC.mkRefCounter1`

File: src/Database/LSMTree/Internal/Merge.hs [hunk 2]
```diff
@@ -113,8 +131,12 @@ complete Merge{..} = do
     readMutVar mergeLastStepResult >>= \case
       MergeInProgress -> error "complete: Merge is not yet completed!"
       MergeComplete -> do
+        -- Since access to a merge /should/ be sequentialised, we can assume
+        -- that the ref count has not changed between this read and the use of
+        -- fromMutable.
+        n <- RC.readRefCount mergeRefCounter
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/Merge.hs:137]: Is it definately sequentialised? What about duplicating or closing other table handles that share this merging run?

Do we need to copy the ref count value? How about reusing/sharing the mutable refcount itself? Then it's safe wrt other threads modifying the count.

**Comment by jorisdral** [comment]
Closing in favour of a new PR

File: src/Database/LSMTree/Internal/Merge.hs (hunk 1 above)
Comment (jorisdral) [reply src/Database/LSMTree/Internal/Merge.hs:88]: Addressed in #423

File: src/Database/LSMTree/Internal/Merge.hs (hunk 2 above)
Comment (jorisdral) [reply src/Database/LSMTree/Internal/Merge.hs:137]: Addressed in #423

**Pull request closed without merging**